File: Codigo_trabalho/Veiculo.cpp
#include "Veiculo.h"

Veiculo::Veiculo()
{
}

Veiculo::Veiculo(int ID) : id{ID}
{
    quantidade_rodas = 4;
    distancia_percorrida = 0;
    combustivel = 0;
    valor_venda = 100;
    ipva = rand() % 2;
    for (int i = 0; i < quantidade_rodas; i++)
        rodas[i] = Rodas();

    cout << "veiculo com ID " << id << " foi criado" << endl;
}

int Veiculo::get_quantidade_rodas()
{
    if (ligado == true)
    {
        return this->quantidade_rodas;
    }
}

void Veiculo::mover()
{
    if (ligado == true)
    {
        if (this->combustivel >= 0.55 && this->ipva == true && verificar_todos_pneus() == true)
            this->distancia_percorrida++;
        else
            cout << "veiculo não possui todos os requisitos para moverse";
    }
}

void Veiculo::abastecer(float abastecer)
{
    if (ligado == true)
    {
        this->combustivel += abastecer;
    }
}

void Veiculo::consumir(float consumir)
{
    if (ligado == true)
    {
        this->combustivel -= consumir;
    }
}

bool Veiculo::verificar_todos_pneus()
{
    bool result;
    if (ligado == true)
    {
        for (int i = 0; i < this->quantidade_rodas; i++)
            result = this->rodas[i].get_calibragem();
        return result;
    }
}
void Veiculo::calibrar(int id_pneu)
{
    if (ligado == true)
    {
        this->rodas[id_pneu].set_calibragem(true);
    }
}

void Veiculo::esvaziar(int id_pneu)
{
    if (ligado == true)
    {
        this->rodas[id_pneu].set_calibragem(false);
    }
}

void Veiculo::imprimir_veiculo()
{
    if (ligado == true)
    {
        int i;
        for (i = 0; i < quantidade_rodas; i++)
            cout << "     ";
        cout << "    ____" << endl;
        for (i = 0; i < quantidade_rodas; i++)
            cout << "     ";
        cout << " __/ |_ \\._" << endl;
        for (i = 0; i < quantidade_rodas; i++)
            cout << "     ";
        cout << "|  _     _``-." << endl;
        for (i = 0; i < quantidade_rodas; i++)
            cout << "     ";
        for (i = 0; i < quantidade_rodas; i++)
            cout << "     ";
        cout << "'-(_)---(_)--'" << endl;
        for (i = 0; i < quantidade_rodas; i++)
            cout << '---------------------' << endl;
    }
}

void Veiculo::mostrar_dados()
{
    if (ligado = true)
    {
        cout << "id: " << this->id << endl;
        cout << "valor de venda: " << this->valor_venda << endl;
        cout << "distancia percorrida: " << this->distancia_percorrida << endl;
        cout << "combustivel: " << this->combustivel << endl;
        cout << "ipva: " << this->ipva << endl;
        cout << "quantidade de rodas: " << this->quantidade_rodas << endl;
        cout << "Rodas: " << this->rodas << endl;
        for (int i = 0; i < this->quantidade_rodas; i++)
        {
            cout << "Roda " << i << ": " << this->rodas[i].get_calibragem() << endl;
        }
    }
}

File: POO/C++/Trabalho_POO_parte_2/Veiculo.py
import Rodas
import random

class Veiculo:
    quantidade_rodas = 4

    def __init__(self):
        self.existe = False

    def __init__(self, ID):
        self.existe = True
        self.id = ID
        self.distancia_percorrida = 0
        self.desenho = "    ____\n __/ |_ \\._\n|  _     _``-.\n-(_)---(_)--\n"
        self.combustivel = 1
        self.valor_venda = 100
        self.ipva = bool(random.randint(0, 1))
        # AVISO PRECISA DE AJUSTE
        # self.rodas = Rodas() for _ in range(self.quantidade_rodas)
        # 
        print(f"Veiculo com ID {self.id} foi inicializado na pista")

    def get_existe(self):
        return self.existe

    def get_quantidade_rodas(self):
        return self.quantidade_rodas

    def mover(self):
        if self.combustivel >= 0.55 and self.ipva and self.verificar_todos_pneus():
            self.consumir(0.55)
            self.distancia_percorrida += 1
        else:
            if self.combustivel < 0.55:
                print(f"Veiculo ID {self.id} nao possui combustivel suficiente")
            if self.ipva:
                print(f"Veiculo ID {self.id} nao possui o ipva pago")
            if self.verificar_todos_pneus():
                print(f"Veiculo ID {self.id} nao possui todos os pneus calibrados")

    def abastecer(self, abastecer):
        self.combustivel += abastecer
        print(f"Tanque de combustivel esta agora com {self.combustivel} de combustivel")

    def consumir(self, consumir):
        if self.combustivel > 0:
            self.combustivel -= consumir
            print(f"Tanque de combustivel esta agora com {self.combustivel} de combustivel")
        else:
            print("Sem gasolina para consumir")
        if self.combustivel < 0:
            self.combustivel = 0

    def verificar_todos_pneus(self):
        result = True
        for roda in self.rodas:
            if not roda.get_calibragem():
                result = False
        return result

    def calibrar(self, id_pneu):
        self.rodas[id_pneu].set_calibragem(True)
        print(f"Pneu: {id_pneu} -> calibrado")

    def esvaziar(self, id_pneu):
        self.rodas[id_pneu].set_calibragem(False)
        print(f"Pneu: {id_pneu} -> esvaziado")

    def imprimir_veiculo(self):
        for _ in range(self.distancia_percorrida):
            print("     ", end="")
        print("    ____")
        for _ in range(self.distancia_percorrida):
            print("     ", end="")
        print(" __/ |_ \\._")
        for _ in range(self.distancia_percorrida):
            print("     ", end="")
        print("|  _     _``-.")
        for _ in range(self.distancia_percorrida):
            print("     ", end="")
        print("-(_)---(_)--\n")

    def mostrar_dados(self):
        print(f"\nID: {self.id}")
        print("IPVA: ", end="")
        if self.ipva:
            print("Sim")
        else:
            print("Nao")
        print(f"Combustivel: {self.combustivel}")
        print(f"Valor de venda: {self.valor_venda}")
        print(f"Distancia percorrida: {self.distancia_percorrida}")
        print(f"Quantidade de rodas: {self.quantidade_rodas}")
        for i in range(self.quantidade_rodas):
            print(f"Roda {i}: ", end="")
            if self.rodas[i].get_calibragem():
                print("calibrada")
            else:
                print("decalibrada")
